keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,565 @@
+
+### Risks and Mitigations
+
+Jobs have a known issue in which more than one Pod can be started even if
+parallelism and completion are set to 1 ([reference]). In the case of indexed
+Jobs, this translates to more than one Pod having the same index.
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### JobSpec API
+
+The JobSpec gets the field `completionMode` to control whether the Job should	
+be treated as an Indexed Job.	
+
+```golang	
+type JobSpec struct {	
+  ...	
+  // CompletionMode specifies how Pod completions are tracked. It can be	
+  // `NonIndexed` (default) or `Indexed`.	
+  // `NonIndexed` means that each Pod completion is homologous to each other.	
+  // The Job is considered complete when there have been .spec.completions	
+  // successful completions.	
+  // `Indexed` means that each Pod completion needs to be tracked individually;	
+  // each Pods gets an associated completion index, which is available in the
+  // annotation	`batch.alpha.kubernetes.io/job-completion-index`.	
+  // The Job is considered complete when there is one successful Pod for each	
+  // index in the range 0 to (.spec.completions - 1).	
+  // When value is `Indexed`, .spec.completions must have a non-zero positive	
+  // value and `.spec.parallelism` must be less than or equal to 10^6.
+  //
+  // More completion modes can be added in the future. If a Job controller
+  // observes a mode that it doesn't recognize, it manages the Job as in
+  // `NonIndexed`.
+  CompletionMode string	
+}	
+```
+
+As the comment describes, when `.spec.completionMode = "Indexed"`:
+
+- The `.spec.completions` must be a non-zero positive value. This is to trigger
+  Jobmanagement strategy for *fixed completion count*. That is, `Indexed` mode
+  cannot	be used for work queue patterns.	
+- The `.spec.parallelism` must be up to `10^6`. This is to guarantee that we
+  can keep track of completions per-index in the Job status in the future.
+
+### Pod detail
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
+Indexed Jobs get the annotation `batch.alpha.kubernetes.io/job-completion-index`
+with a value equal to its completion index. The annotation is immutable.
+
+The annotation can be accessed through the downward API as a file or environment
+variable.
+
+For user convenience, the Job controller adds the completion index as an
+environment variable through the downward API. That is, the Job controller
+creates Pods like so:
+
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+    - name: test-container
+      env:
+        - name: JOB_COMPLETION_INDEX
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.annotations['batch.alpha.kubernetes.io/job-completion-index'] 
```

Comment:
Isn't this part of the Job's pod template, as User Story 1 shows?

Or it's something that automatically created for the users, without them specifying it in the pod template?

Reply:
It's automatically added at Pod creation, but user can also specify their own. See updated note.